File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        top,bottom=0,len(matrix)-1
        left,right=0,len(matrix[0])-1
        if not matrix: return []
        res=[]
        while top<=bottom and left<=right:
            #traverse the first row
            for i in range(left,right+1):
                res.append(matrix[top][i])
            top+=1
            #last column
            for i in range(top,bottom+1):
                res.append(matrix[i][right])
            right-=1
            if top <=bottom:
            #lastrow in reverse
                for i in range(right,left-1,-1):
                    res.append(matrix[bottom][i])
                bottom-=1
            #column form bottom to top
            if left<=right:
                for i in range(bottom,top-1,-1):
                    res.append(matrix[i][left])
                left+=1
        return res
            

File: 0054-spiral-matrix/0054-spiral-matrix.c
int* spiralOrder(int** matrix, int matrixSize, int* matrixColSize, int* returnSize) {

    int rows = matrixSize;
    int cols = matrixColSize[0];

    static int result[1000];

    int top = 0;
    int bottom = rows - 1;
    int left = 0;
    int right = cols - 1;

    int k = 0;

    while(top <= bottom && left <= right) {

        // Left to Right
        for(int i = left; i <= right; i++) {
            result[k++] = matrix[top][i];
        }
        top++;

        // Top to Bottom
        for(int i = top; i <= bottom; i++) {
            result[k++] = matrix[i][right];
        }
        right--;

        // Right to Left
        if(top <= bottom) {
            for(int i = right; i >= left; i--) {
                result[k++] = matrix[bottom][i];
            }
            bottom--;
        }

        // Bottom to Top
        if(left <= right) {
            for(int i = bottom; i >= top; i--) {
                result[k++] = matrix[i][left];
            }
            left++;
        }
    }

    *returnSize = k;

    return result;
}

// Synced seamlessly with LeetHub Pro
// Pro features: https://bit.ly/leethubpro | Free version: https://bit.ly/leethubv4
// Get it here: https://chromewebstore.google.com/detail/leethub-v4/bcilpkkbokcopmabingnndookdogmbna